src/v/cloud_roles/azure_aks_refresh_impl.cc

```diff
@@ -0,0 +1,193 @@
+    }(),
+    std::move(region),
+    as,
+    retry_params)
+  , client_id_{load_from_env(env_var_azure_client_id)}
+  , tenant_id_{load_from_env(env_var_azure_tenant_id)}
+  , federated_token_file_{load_from_env(env_var_azure_federated_token_file)} {}
+
+auto azure_aks_refresh_impl::print(std::ostream& os) const -> std::ostream& {
+    fmt::print(os, "azure_aks_refresh_impl{{address:{}}}", address());
+    return os;
+}
+
+auto azure_aks_refresh_impl::fetch_credentials() -> ss::future<api_response> {
+    // https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-client-creds-grant-flow#third-case-access-token-request-with-a-federated-credential
+    // load the encoded jwt from the token file (its content is rotated by k8s),
+    // compose:
+    // POST /{tenant_id_}/oauth2/v2.0/token HTTP/1.1               // Line
+    // breaks for clarity Host: {authority_host_}:443 Content-Type:
+    // application/x-www-form-urlencoded
+    //
+    // scope=https%3A%2F%2Fgraph.microsoft.com%2F.default
+    // &client_id={client_id_}
+    // &client_assertion_type=urn%3Aietf%3Aparams%3Aoauth%3Aclient-assertion-type%3Ajwt-bearer
+    // &client_assertion={content(federated_token_file_)}
+    // &grant_type=client_credentials
+
+    auto jwt_token_fut = co_await ss::coroutine::as_future(
+      read_fully_to_string({federated_token_file_}));
+    if (jwt_token_fut.failed()) {
+        vlog(
+          clrl_log.error,
+          "failed to read OIDC pod token from file {}",
+          federated_token_file_);
+        co_return ss::coroutine::exception(jwt_token_fut.get_exception());
+    }
+
+    auto jwt_token = std::string{jwt_token_fut.get()};
+    boost::trim(jwt_token);
+
+    // setup header for www-form-urlencoded body
+    auto access_token_req
+      = boost::beast::http::request<boost::beast::http::string_body>{};
+    access_token_req.method(boost::beast::http::verb::post);
+    access_token_req.target(fmt::format("/{}/oauth2/v2.0/token", tenant_id_));
+    access_token_req.set(
+      boost::beast::http::field::host,
+      fmt::format("{}:{}", address().host(), address().port())),
+      access_token_req.set(
+        boost::beast::http::field::user_agent, "redpanda.vectorized.io");
+    access_token_req.set(
+      boost::beast::http::field::content_type,
+      "application/x-www-form-urlencoded");
+
+    // compose the body of the request with the informations retrieved from the
+    // env variables
+    auto body = ssx::sformat(
+      "scope={}"
+      "&client_id={}"
+      "&client_assertion_type=urn%3Aietf%3Aparams%3Aoauth%3Aclient-assertion-"
+      "type%3Ajwt-bearer"
+      "&client_assertion={}"
+      "&grant_type=client_credentials",
+      uri_encode("https://storage.azure.com/.default", true),
+      client_id_,
+      jwt_token);
+
+    // TODO check
+    // AKS requires a TLS enabled client by default, but in test mode where we
+    // use the http imposter, we need to use a simple client.
```

Comment:
i don't understand this comment. what changes if AKS requires a TLS enabled by default or it doesn't? is this a follow-up PR or ticket that needs to be filed?

Reply:
it's related to the fact that we run azurite and the http_imposter on http, while the real service would require https.